File: android/src/main/java/com/parityblur/WindowBlurContext.kt
package com.parityblur

import android.view.Choreographer
import android.view.View
import android.view.ViewTreeObserver
import java.util.Collections
import java.util.WeakHashMap

/**
 * Per-window active-view registry + capture scheduling (plan §4, §14.3, §14.4, §20, §21, §39).
 * One instance per native window (keyed by the window's root view in [BlurEngine]) --
 * NEVER one scheduler per [ParityBlurView] (plan §45.6).
 *
 * STATIC mode: capture requests (refresh(), layout/prop changes) coalesce to ONE capture per
 * view at the next frame boundary via [Choreographer.postFrameCallback] -- never synchronously
 * inside JS-command dispatch, never invalidating from preDraw (plan §45.8).
 *
 * LIVE mode (Milestone 6): ONE [ViewTreeObserver.OnPreDrawListener] per window, installed only
 * while at least one live view is registered and removed when the last leaves (plan §24: zero
 * scheduler work when no live blur is active). Each pre-draw tick:
 *   1. filters views by the visibility heuristic (plan §24),
 *   2. applies the per-view maxFps throttle (plan §21: respect elapsed time),
 *   3. recaptures INSIDE the in-flight frame -- the re-recorded RenderNode propagates its own
 *      damage, so no invalidate() is issued from preDraw (M0-proven; plan §45).
 * Captures here are synchronous software captures (see [ParityBlurView.performLiveCapture]),
 * so backpressure (plan §22) is inherent: at most one capture per view per frame, and a slow
 * frame simply delays the next tick -- no queue can form.
 */
class WindowBlurContext(private val rootView: View) {

  /** All currently-attached ParityBlurView instances in this window (weak, for bookkeeping). */
  private val activeViews: MutableSet<ParityBlurView> =
    Collections.newSetFromMap(WeakHashMap())

  private val pendingCaptures = LinkedHashSet<ParityBlurView>()
  private var frameCallbackScheduled = false

  private val frameCallback = Choreographer.FrameCallback {
    frameCallbackScheduled = false
    // Snapshot-and-clear before iterating: a capture triggered from within this callback
    // schedules a NEW frame instead of mutating the set being iterated right now.
    val batch = pendingCaptures.toList()
    pendingCaptures.clear()
    for (view in batch) {
      if (view.isAttachedToWindow) {
        view.performScheduledCapture()
      }
    }
  }

  // ------------------------------------------------------------------- live (Milestone 6)

  private val liveViews: MutableSet<ParityBlurView> =
    Collections.newSetFromMap(WeakHashMap())

  /**
   * Static views whose window position is watched each frame (plan §18). Separate from
   * [liveViews]: these do NOT recapture per frame, they only re-request a capture when their
   * position in the window actually changes.
   */
  private val geometryWatched: MutableSet<ParityBlurView> =
    Collections.newSetFromMap(WeakHashMap())

  private var preDrawInstalled = false

  private val preDrawListener = ViewTreeObserver.OnPreDrawListener {
    val now = System.nanoTime()
    for (view in liveViews) {
      if (view.isLiveEligible(now)) {
        view.performLiveCapture(now)
      }
    }
    // Geometry watch (plan §18): the only trigger that sees an ancestor TRANSFORM -- onSizeChanged
    // and onLayout both miss it, so without this a static backdrop inside a transform-animated
    // sheet host freezes whatever partial band it captured mid-animation. Each view costs one
    // getLocationInWindow + two int compares; a changed position only REQUESTS a capture, which the
    // Choreographer path then coalesces as usual.
    for (view in geometryWatched) {
      view.checkWindowGeometry()
    }
    true
  }

  /**
   * Install/uninstall the shared pre-draw listener to match demand (plan §24): it runs only while
   * this window actually has live views or watched static views. NOTE: plan §24's "zero scheduler
   * work when no live blur is active" now reads "no live blur AND no attached static blur" -- a
   * static view must be watched to notice an ancestor transform. The per-frame cost of a watched
   * static view is a position compare, not a capture.
   */
  private fun syncPreDrawListener() {
    val wanted = liveViews.isNotEmpty() || geometryWatched.isNotEmpty()
    if (wanted && !preDrawInstalled) {
      val vto = rootView.viewTreeObserver
      if (vto.isAlive) {
        vto.addOnPreDrawListener(preDrawListener)
        preDrawInstalled = true
        ParityBlurDebug.log { "scheduler-install window=${System.identityHashCode(rootView)}" }
      }
    } else if (!wanted && preDrawInstalled) {
      val vto = rootView.viewTreeObserver
      if (vto.isAlive) vto.removeOnPreDrawListener(preDrawListener)
      preDrawInstalled = false
      ParityBlurDebug.log { "scheduler-uninstall window=${System.identityHashCode(rootView)}" }
    }
  }

  fun registerGeometryWatch(view: ParityBlurView) {
    geometryWatched.add(view)
    syncPreDrawListener()
  }

  fun unregisterGeometryWatch(view: ParityBlurView) {
    geometryWatched.remove(view)
    syncPreDrawListener()
  }

  fun register(view: ParityBlurView) {
    activeViews.add(view)
  }

  fun unregister(view: ParityBlurView) {
    activeViews.remove(view)
    pendingCaptures.remove(view)
    unregisterLive(view)
    unregisterGeometryWatch(view)
  }

  fun registerLive(view: ParityBlurView) {
    liveViews.add(view)
    syncPreDrawListener()
  }

  fun unregisterLive(view: ParityBlurView) {
    liveViews.remove(view)
    syncPreDrawListener()
  }

  /**
   * Coalesce a capture request for [view] to the next valid frame boundary. Repeated calls
   * before the frame fires collapse into one capture (plan §14.4/§20).
   */
  fun scheduleCapture(view: ParityBlurView) {
    pendingCaptures.add(view)
    if (!frameCallbackScheduled) {
      frameCallbackScheduled = true
      Choreographer.getInstance().postFrameCallback(frameCallback)
    }
  }

  fun cancelScheduledCapture(view: ParityBlurView) {
    pendingCaptures.remove(view)
  }

  fun isEmpty(): Boolean = activeViews.isEmpty()
}

File: ios/WindowBlurContext.swift
import Foundation
import UIKit

/// Per-window active-view registry + capture scheduling (plan §15.2, §20, §21, §39).
///
/// STATIC mode: capture requests coalesce to ONE capture per view on the next main-runloop
/// turn (plan §29, §45.8).
///
/// LIVE mode (Milestone 6): ONE shared CADisplayLink per window — never one per view (plan
/// §45.6) — created when the first live view registers and torn down when the last leaves
/// (plan §24/§27: zero frame work while no live blur is visible). Each tick filters by the
/// visibility heuristic and each view's own maxFps throttle + in-flight backpressure
/// (plan §21/§22: at most 1 in-flight GPU pass per view, latest-wins, stale frames dropped).
/// The link pauses while the app is inactive (plan §21 app/scene lifecycle pause).
final class WindowBlurContext {

  /// Attached ParityBlur core views in this window (weak bookkeeping).
  private let activeViews = NSHashTable<ParityBlurCoreView>.weakObjects()

  private var pendingCaptures = NSHashTable<ParityBlurCoreView>.weakObjects()
  private var flushScheduled = false

  // ------------------------------------------------------------------- live (Milestone 6)

  private let liveViews = NSHashTable<ParityBlurCoreView>.weakObjects()
  private var displayLink: CADisplayLink?
  private var lifecycleObservers: [NSObjectProtocol] = []

  // ---------------------------------------------------------------- settle poll (plan §18)

  /// Static views whose capture would currently be CLAMPED by the target bounds because an
  /// ancestor transform has moved them partly outside the window. Polled until their geometry
  /// stops moving; see `ParityBlurCoreView.performScheduledCapture`.
  private let settleViews = NSHashTable<ParityBlurCoreView>.weakObjects()
  private var settleLink: CADisplayLink?

  /// CADisplayLink retains its target; this proxy breaks the cycle so the context (and its
  /// window) can deallocate naturally.
  private final class LinkProxy {
    weak var context: WindowBlurContext?
    init(_ c: WindowBlurContext) { context = c }
    @objc func tick(_ link: CADisplayLink) { context?.liveTick(link) }
    @objc func settleTick(_ link: CADisplayLink) { context?.settleTick(link) }
  }

  deinit {
    displayLink?.invalidate()
    settleLink?.invalidate()
    for o in lifecycleObservers { NotificationCenter.default.removeObserver(o) }
  }

  /**
   Poll [view] on each frame until its window geometry settles (plan §18).

   `layoutSubviews` only fires on a `bounds.size` change, so an ancestor TRANSFORM -- what every
   sheet/modal transition animates -- is invisible to every other trigger, while
   `convert(bounds, to: window)` (which the capture plan is built from) does reflect it. Without
   this, a fullscreen backdrop inside a transform-animated host captures a clamped band mid-
   animation and nothing ever re-captures it.

   Deliberately NOT a permanently-installed link: unlike Android's pre-draw listener (which only
   runs when a frame is already being drawn, and is therefore free at idle), a CADisplayLink ticks
   every frame regardless. This one is armed only while a clamped-and-moving view exists and is
   torn down the moment geometry settles, so a settled window keeps ZERO links -- preserving the
   M7 teardown guarantee (docs/HARDENING_REPORT.md).
   */
  func scheduleSettlePoll(_ view: ParityBlurCoreView) {
    settleViews.add(view)
    guard settleLink == nil else { return }
    let link = CADisplayLink(target: LinkProxy(self), selector: #selector(LinkProxy.settleTick(_:)))
    link.add(to: .main, forMode: .common)
    settleLink = link
    ParityBlurDebug.log("settle-install window=\(ObjectIdentifier(self))")
  }

  private func settleTick(_ link: CADisplayLink) {
    let batch = settleViews.allObjects
    settleViews.removeAllObjects()
    // checkWindowGeometry requests a capture when the origin moved, and reports whether the view is
    // still clamped. Only still-clamped views stay in the poll, so the link tears itself down as
    // soon as the last backdrop settles fully inside the window.
    for v in batch where v.window != nil {
      if v.checkWindowGeometry() { settleViews.add(v) }
    }
    if settleViews.count == 0 {
      settleLink?.invalidate()
      settleLink = nil
      ParityBlurDebug.log("settle-uninstall window=\(ObjectIdentifier(self))")
    }
  }

  func register(_ view: ParityBlurCoreView) {
    activeViews.add(view)
  }

  func unregister(_ view: ParityBlurCoreView) {
    activeViews.remove(view)
    pendingCaptures.remove(view)
    settleViews.remove(view)
    if settleViews.count == 0, settleLink != nil {
      settleLink?.invalidate()
      settleLink = nil
    }
    unregisterLive(view)
  }

  func registerLive(_ view: ParityBlurCoreView) {
    liveViews.add(view)
    ensureDisplayLink()
  }

  func unregisterLive(_ view: ParityBlurCoreView) {
    liveViews.remove(view)
    if liveViews.count == 0 {
      displayLink?.invalidate()
      displayLink = nil
      ParityBlurDebug.log("scheduler-uninstall window=\(ObjectIdentifier(self))")
    }
  }

  private func ensureDisplayLink() {
    guard displayLink == nil else { return }
    let link = CADisplayLink(target: LinkProxy(self), selector: #selector(LinkProxy.tick(_:)))
    // Cap the link at the highest maxFps among live views; per-view throttles refine below it.
    let cap = Float(liveViews.allObjects.map(\.maxFps).max() ?? 30)
    link.preferredFrameRateRange = CAFrameRateRange(minimum: 10, maximum: cap, preferred: cap)
    link.add(to: .main, forMode: .common)
    displayLink = link
    ParityBlurDebug.log("scheduler-install window=\(ObjectIdentifier(self))")

    if lifecycleObservers.isEmpty {
      let nc = NotificationCenter.default
      lifecycleObservers.append(nc.addObserver(
        forName: UIApplication.willResignActiveNotification, object: nil, queue: .main
      ) { [weak self] _ in self?.displayLink?.isPaused = true })
      lifecycleObservers.append(nc.addObserver(
        forName: UIApplication.didBecomeActiveNotification, object: nil, queue: .main
      ) { [weak self] _ in self?.displayLink?.isPaused = false })
    }
  }

  private func liveTick(_ link: CADisplayLink) {
    let now = link.timestamp
    for view in liveViews.allObjects where view.isLiveEligible(now: now) {
      view.performLiveTick(now: now)
    }
  }

  /// Coalesce a static capture request to the next main-runloop turn (plan §20).
  func scheduleCapture(_ view: ParityBlurCoreView) {
    pendingCaptures.add(view)
    guard !flushScheduled else { return }
    flushScheduled = true
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      self.flushScheduled = false
      let batch = self.pendingCaptures.allObjects
      self.pendingCaptures.removeAllObjects()
      for v in batch where v.window != nil {
        v.performScheduledCapture()
      }
    }
  }

  var isEmpty: Bool { activeViews.count == 0 }
}
